include/zephyr/sys/util.h

```diff
@@ -783,6 +783,39 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
 	mem_xor_n(dst, src1, src2, 16);
 }
 
+/**
+ * @brief Returns the number of bits set in a value
+ *
+ * @param value The value to count number of bits set of
+ * @param len The number of octets in @p value
+ */
+static inline size_t zephyr_count_bits(const void *value, size_t len)
+{
+	size_t cnt = 0U;
+	size_t i = 0;
+
+#ifdef POPCOUNT
+	for (; i < len / sizeof(unsigned int); i++) {
+		unsigned int val = ((unsigned int *)value)[i];
+
+		cnt += POPCOUNT(val);
+	}
+	i *= sizeof(unsigned int); /* convert to a uint8_t index for the remainder (if any) */
+#endif
```

Comment:
POPCOUNT works very well for `unsigned int` types for both GCC and Clang. There are additional POPCOUNT functions for other sizes, but not sure if they provide any meaningful optimizations. 

It is, however (from my benchmark tests) very important to provide a unsigned int to GCC; there was a factor of more than 10000 in terms of speed on whether I provided a `unsigned int` or a `uint8_t` to POPCOUNT, so this version is the fastest I've been able to produce which handles any size, but being extremely fast if POPCOUNT has been defined.